src/Azure.Deployments.Extensibility.Providers.Graph/GraphProvider.cs

```diff
@@ -0,0 +1,252 @@
+                //2.1.1 Get POST uri and POST to create resource
+                var postUri = GeneratePostUri(resource.Type, resource.Properties);
+                response = await _graphHttpClient.PostAsync(postUri, properties, graphToken.GetString()!, cancellationToken);
+            }
+            else
+            {
+                // 2.2 Update resource if successfully get id
+                // 2.2.1 Get PATCH uri and PATCH to update resource
+                var patchUri = GeneratePatchUri(resource.Type, id);
+                response = await _graphHttpClient.PatchAsync(patchUri, properties, graphToken.GetString()!, cancellationToken);
+            }
+
+            // Return properties in request if none returned in response
+            var content = HandleHttpResponse(response);
+            var resultProperties = response.StatusCode == HttpStatusCode.NoContent ? resource.Properties : JsonSerializer.Deserialize<JsonElement>(content);
+
+            return new ExtensibilityOperationSuccessResponse(
+                request.Resource with
+                {
+                    Properties = resultProperties
+                }
+            );
+        }
+
+        /*
+         * Returns actual id from response when
+         * Response returns successfully
+         * AND 
+         *  there is an attribute "id" in response (for users)
+         *  OR "value" in response is not empty (for others)
+         */
+        public static string GetIdIfExists(HttpResponseMessage response, string uri)
+        {
+            var content = response.Content.ReadAsStringAsync().Result;
+            var id = "";
+
+            if (response.IsSuccessStatusCode)
+            {
+                var contentElement = JsonSerializer.Deserialize<JsonElement>(content);
+
+                if (contentElement.TryGetProperty("id", out JsonElement idElement))
+                {
+                    id = idElement.ToString();
+                }
+                else
+                {
+                    var value = contentElement.GetProperty("value");
+
+                    if (value.GetArrayLength() > 0)
+                    {
+                        var firstValue = value.EnumerateArray().First();
+                        id = firstValue.GetProperty("id").ToString();
+                    }
+                }
+
+            }
+            else if (response.StatusCode != HttpStatusCode.NotFound || !uri.StartsWith("users"))
+            {
+                throw new GraphHttpException((int)response.StatusCode, content);
+            }
+
+            return id;
+        }
+
+        public static string GenerateGetUri(string resourceType, JsonElement properties)
+        {
+            var typesAndVersion = resourceType.Split('@');
+            var types = typesAndVersion[0].Split('/');
+            var uri = types[1];
+
+            // TODO: Use `uniqueName` when ready
+            if (types.Length == 2 && uri == "users")
```

Comment:
Eventually we want to use `uniqueName` when its ready. Using `name` for `users` here to demonstrate that `name` can be used as key. For this POC, I think we can also use `displayName` for `users` as well for consistency and making the provider type-generic. Open to suggestions

Reply:
Service principals should use application name